tests/cosim/emulator/src/spike_event.cc

```diff
@@ -0,0 +1,241 @@
+    // xx0000 <- x
+    // xx0001 <- f
+    // xx0010 <- vreg
+    // xx0011 <- vec
+    // xx0100 <- csr
+
+    if ((write_idx & 0xf) == 0b0000) {// scalar rf
+      uint64_t rd_should_be_bits = proc.get_state()->XPR[rd_idx];
+      if (rd_new_bits != rd_should_be_bits) {
+        rd_new_bits = rd_should_be_bits;
+        is_rd_written = true;
+        LOG(INFO) << fmt::format("Log Spike {:08X} with scalar rf change: x[{}] from {:08X} to {:08X}", pc, rd_idx, rd_old_bits, rd_new_bits);
+      }
+    }
+  }
+
+  for (auto mem_write: state->log_mem_write) {
+    uint64_t address = std::get<0>(mem_write);
+    if (address != target_mem) {
+      LOG(FATAL) << fmt::format("Error! spike detect mem_write at= {:08X}; target mem = {:08X}", address, target_mem);
+    }
+    uint64_t value = std::get<1>(mem_write);
+    // Byte size_bytes
+    uint8_t size_by_byte = std::get<2>(mem_write);
+    LOG(INFO) << fmt::format("spike detect mem write {:08X} on mem:{:08X} with size={}byte", value, address, size_by_byte);
+    mem_access_record.all_writes[address] = {.size_by_byte = size_by_byte, .val = value};
+  }
+  // since log_mem_read doesn't record mem data, we need to load manually
+  for (auto mem_read: state->log_mem_read) {
+    uint64_t address = std::get<0>(mem_read);
+    if (address != target_mem) {
+      LOG(FATAL) << fmt::format("Error! spike detect mem_read at= {:08X}; target mem = {:08X}", address, target_mem);
+    }
+    // Byte size_bytes
+    uint8_t size_by_byte = std::get<2>(mem_read);
+    uint64_t value = 0;
+    // record mem target
+    for (int i = 0; i < size_by_byte; ++i) {
+      value += (uint64_t) impl->load(address + i) << (i * 8);
+    }
+    LOG(INFO) << fmt::format("spike detect mem read {:08X} on mem:{:08X} with size={}byte", value, address, size_by_byte);
+    mem_access_record.all_reads[address] = {.size_by_byte = size_by_byte, .val = value};
+  }
+
+  // record root page table
+  if (satp_mode == 0x8 && block.addr == -1) {
+    uint64_t root_addr = satp_ppn << 12;
+    for (int i = 0; i < 8; i++) {
+      uint64_t data = 0;
+      //scan 8 bytes to data
+      for (int j = 0; j < 8; ++j) {
+        data += (uint64_t) impl->load(root_addr + j + i * 8) << (j * 8);
+      }
+      block.blocks[i] = data;
+      block.addr = root_addr;
+      block.remaining = true;
+    }
+  }
+
+  state->log_reg_write.clear();
+  state->log_mem_read.clear();
+  state->log_mem_write.clear();
+}
+
+SpikeEvent::SpikeEvent(processor_t &proc, insn_fetch_t &fetch, VBridgeImpl *impl) : proc(proc), impl(impl) {
+  auto &xr = proc.get_state()->XPR;
+  pc = proc.get_state()->pc;
+  inst_bits = fetch.insn.bits();
+  target_mem = -1;
+  is_committed = false;// j insn should be committed immediately cause it doesn't have wb stage.
+
+  // extension depending parameter
+  is_compress = false;
+  if (fetch.insn.length() == 2) {
+    is_compress = true;
+  }
+  if (!is_compress) {
+    rs1_bits = xr[fetch.insn.rs1()];
```

Comment:
I don't think all of this should be necessary... this reimplements details of execution. Why?